lib/core/macrocache.nim

```diff
@@ -7,38 +7,199 @@
+  ## Returns the value of a counter `c`.
+  runnableExamples:
+    static:
+      let counter = CacheCounter"valTest"
+      # default value is 0
+      assert counter.value == 0
+
+      inc counter
+      assert counter.value == 1
+
+proc inc*(c: CacheCounter; by = 1) {.magic: "NccInc".} =
+  ## Increments the counter `c` with the value `by`.
+  runnableExamples:
+    static:
+      let counter = CacheCounter"incTest"
+      inc counter
+      assert counter.value == 1
+
+      inc counter, 5
+      assert counter.value == 6
+
+proc add*(s: CacheSeq; value: NimNode) {.magic: "NcsAdd".} =
+  ## Adds `value` to `s`.
+  runnableExamples:
+    import std/macros
+    const mySeq = CacheSeq"addTest"
+    
+    static:
+      mySeq.add(newLit(5))
+      mySeq.add(newLit("hello ic"))
+  
+      assert mySeq.len == 2
+      assert mySeq[1].strVal == "hello ic"
+
+proc incl*(s: CacheSeq; value: NimNode) {.magic: "NcsIncl".} = 
+  ## Adds `value` to `s`.
+  ## 
+  ## .. hint:: This doesn't do anything if `value` is already in `s`.
+  runnableExamples:
+    import std/macros
+    const mySeq = CacheSeq"inclTest"
+    
+    static:
+      mySeq.incl(newLit(5))
+      mySeq.incl(newLit(5))
 
-proc value*(c: CacheCounter): int {.magic: "NccValue".}
-proc inc*(c: CacheCounter; by = 1) {.magic: "NccInc".}
+      # still one element
+      assert mySeq.len == 1
 
-proc add*(s: CacheSeq; value: NimNode) {.magic: "NcsAdd".}
-proc incl*(s: CacheSeq; value: NimNode) {.magic: "NcsIncl".}
-proc len*(s: CacheSeq): int {.magic: "NcsLen".}
-proc `[]`*(s: CacheSeq; i: int): NimNode {.magic: "NcsAt".}
+proc len*(s: CacheSeq): int {.magic: "NcsLen".} =
+  ## Returns the length of `s`.
+  runnableExamples:
+    import std/macros
+
+    const mySeq = CacheSeq"lenTest"
+    static:
+      let val = newLit("helper")
+      mySeq.add(val)
+      assert mySeq.len == 1
+      for _ in 0..<5:
+        mySeq.add(val)
+      
+      assert mySeq.len == 6
+
+proc `[]`*(s: CacheSeq; i: int): NimNode {.magic: "NcsAt".} =
+  ## Returns the `i`th value from `s`.
+  runnableExamples:
+    import std/macros
+
+    const mySeq = CacheSeq"subTest"
+    static:
+      # add 42 values
+      for i in 0..41:
+        mySeq.add(newLit(i + 1))
+      
+      assert mySeq[41].intVal == 42
 
 iterator items*(s: CacheSeq): NimNode =
+  ## Iterates over each item in `s`.
+  runnableExamples:
+    import std/macros
+    const myseq = CacheSeq"itemsTest"
+    
+    static:
+      myseq.add(newLit(5))
+      myseq.add(newLit(42))
+      
+      for val in myseq:
+        # check that all values in `myseq` are int literals
+        assert val.kind == nnkIntLit
+  
   for i in 0 ..< len(s): yield s[i]
 
-proc `[]=`*(t: CacheTable; key: string, value: NimNode) {.magic: "NctPut".}
-  ## 'key' has to be unique!
+proc `[]=`*(t: CacheTable; key: string, value: NimNode) {.magic: "NctPut".} =
+  ## Inserts a `(key, value)` pair into `t`. 
+  ## 
+  ## .. warning:: `key` has to be unique!
```

Comment:
What does "unique" mean here? Unique in the sense of [isolated](https://nim-lang.github.io/Nim/isolation.html)? Or that `key` shouldn't already be in `t`?
And in how far does it "have to" be unique, i.e. what happens otherwise? Nothing? An exception? Undefined behaviour?

Reply:
@konsumlamm well, that warning was before me, but I will of course improve it. If the key is already in the table, then the compiler shows an error which can't be caught (as far as I know) by the macro code. And there's also no way to check if a key is already in the table (without iterating all the values with pairs) :P

Reply:
> there's also no way to check if a key is already in the table

is `hasKey` missing then?

Reply:
@timotheecour I mean it would be nice to have it

Reply:
@konsumlamm I clarified that the compiler will error out if key is already in the table